[Lee I-Shiuan] iP

ish1506 opened the pull request:
# Duke 

>Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke helps you stay on top of your work. It's:

* text-based
* easy to learn
* ~~FAST~~ *SUPER FAST* to use 😃

All you need to do is:

1. download it from [here](https://github.com/ish1506/ip/releases/tag/A-Jar).
2. use the `java -jar duke.jar` command to run the program.
3. add your tasks. 
4. let it manage your tasks for you!

And it's **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Duke {
    public static void main(String[] args) {
        new Duke().run();
    }
}
```

Commit:
Add Gradle support

ish1506 committed:
Implement Skeletal Duke chatbot

ish1506 committed:
Implement add and list functionality

ish1506 committed:
Create Task class

ish1506 committed:
Implement mark and unmark functionalities

ish1506 committed:
Create Deadline, Event and Todo classes

ish1506 committed:
Implement Todo, Deadline and Event task management in Duke chatbot

ish1506 committed:
Write testcases for adding tasks, listing, marking and unmarking

ish1506 committed:
Create DukeException class

ish1506 committed:
Implement error handling for empty Todos and invalid commands

ish1506 committed:
Add class files to .gitignore

ish1506 committed:
Implement delete tasks functionality

ish1506 committed:
Add test inputs for task deletion

ish1506 committed:
Add data directory to .gitignore

ish1506 committed:
Create FileHandler class to read and write tasks data to hard disk

ish1506 committed:
Implement save tasks functionality

ish1506 committed:
Add dates to Deadline class

ish1506 committed:
Merge branch 'branch-Level-7'

ish1506 committed:
Resolve merge conflict with branch-Level-8

ish1506 committed:
Fix bug: deadline dates from hard disk

ish1506 committed:
Update test input and expected output

ish1506 committed:
Rename FileHandler class to Storage

ish1506 committed:
Extract the Tasklist class

ish1506 committed:
Extract the Ui class

ish1506 committed:
Create Command enums

ish1506 committed:
Extract Parser class

ish1506 committed:
Organize classes into packages

ish1506 committed:
Merge branch 'add-gradle-support'

ish1506 committed:
Rename mainClassName in build.gradle

ish1506 committed:
Update README

ish1506 committed:
Organize classes into packages

ish1506 committed:
Rename class name in runtest.bat

ish1506 committed:
Write unit tests for Parser and TaskList classes

ish1506 committed:
Add Javadocs comments

ish1506 committed:
Implement Find functionality in Duke

ish1506 committed:
Merge branch 'branch-A-JavaDoc'

ish1506 committed:
Merge branch 'branch-Level-9'

ish1506 committed:
Fix bug: Incorrect Task string format being written to Storage

ish1506 committed:
Change Javadocs comment to fit Coding Standards

ish1506 committed:
Resolve merge conflicts with branch-A-CodingStandard

ish1506 committed:
Remove Javadocs comments for getters

JasonCP14 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,97 @@
+package duke;
+
+import duke.command.Command;
+import duke.exceptions.DukeException;
+import duke.parser.Parser;
+import duke.storage.Storage;
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.TaskList;
+import duke.task.Todo;
+import duke.ui.Ui;
+
+import java.util.Scanner;
```
Perhaps a better way to write import statements is to order them on:
1. Java Packages, then Custom Packages
2. Lexical Order

JasonCP14 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,124 @@
+     */
+    public void parse() throws DukeException {
+        List<String> listOfInputs = Arrays.asList(input.split(" "));
+        if (input.isEmpty() || listOfInputs.isEmpty()) {
```
Don't you think a series of switch case statements will be better than such large number of if else statements?

JasonCP14 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,97 @@
+package duke;
```
Your variable names are intuitive and are following the coding standard. Nicely done!

JasonCP14 reviewed (changes requested):
LGTM, almost. Just a few nits to fix.

isanidiot commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,97 @@
+
+                switch (command) {
+                case BYE:
+                    // Write tasks data to storage before terminating program
```
Perhaps you can do this after each time the list is updated?

isanidiot commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,5 @@
+package duke.command;
+
```
I like how the enum has its own class.

isanidiot reviewed (commented):
Keep up the good work in week 5! I know GUI is hard...

ish1506 committed:
Create GUI using JavaFX

ish1506 committed:
Refactor with FXML

ish1506 committed:
Reorder import statements

ish1506 committed:
Merge branch 'branch-Level-10'

ish1506 committed:
Refactor code to improve quality

Parser::parse has deeply nested if-else statements which are hard to read.

By using a switch statement and guard clause, the code has fewer indentations and is more readable.
An integerParser method was extracted to prevent code repetition in Parser::parse.

ish1506 committed:
Add error messages for TaskList index out-of-bounds errors

ish1506 committed:
Add assertions in the Duke and CommandHandler classes

The assertions in CommandHandler::execute and Duke::getResponse ensure
that a response message is always returned.

ish1506 committed:
Merge pull request #2 from ish1506/branch-A-Assertions

branch-A-Assertions

ish1506 committed:
Merge branch 'master' into branch-A-CodeQuality

ish1506 committed:
Merge pull request #3 from ish1506/branch-A-CodeQuality

Refactor code to improve quality

ish1506 committed:
Add shortcuts for all commands

ish1506 committed:
Merge branch 'branch-C-FriendlierSyntax'

ish1506 committed:
Delete text-ui-test files

ish1506 committed:
Configure JavaFX version in .fxml files

ish1506 committed:
Add fallthrough comments for switch-case statement

ish1506 committed:
Fix bug: throw DukeException when TaskList is empty

ish1506 committed:
Add title to GUI

ish1506 committed:
Update User Guide and add Ui screenshot to /docs

ish1506 committed:
Update User Guide and add Ui screenshot to /docs

ish1506 committed:
Update User Guide and add Ui screenshot to /docs

ish1506 committed:
Configure Checkstyle

ish1506 committed:
Fix coding style errors
